### PR TITLE
Add GH Actions workflow that automates Changelog PR

### PR DESCRIPTION
**Pull request by joshuacwnewton**
#### GitHub

- [X] I've given this PR a concise, self-descriptive, and meaningful title
- [X] I've linked relevant issues in the PR body
- [X] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [X] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [SCT's internal developer documentation](https://github.com/neuropoly/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

Currently, our [release process](https://github.com/neuropoly/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release) involves manually running the [`changelog`](https://github.com/neuropoly/changelog) utility then manually opening a PR with the changes.

To automate this process, this PR adds a GH Actions workflow using [`create-pull-request`](https://github.com/peter-evans/create-pull-request) that you can manually trigger with a button press:

Screenshot from 2021-03-07 11-52-30

Result: A pull request will be automatically created containing an updated changelog. 

Here are some sample PRs generated by a matching workflow file in a separate repo (https://github.com/joshuacwnewton/test-github-actions/blob/master/.github/workflows/create-changelog-PR.yml):
* https://github.com/joshuacwnewton/test-github-actions/pull/15 (Generated by [this run](https://github.com/joshuacwnewton/test-github-actions/actions/runs/633609751))
* https://github.com/joshuacwnewton/test-github-actions/pull/16 (Generated by [this run](https://github.com/joshuacwnewton/test-github-actions/runs/2279333327?check_suite_focus=true))

When this PR is combined with #3198, then the entire first section of [Creating a new release](https://github.com/neuropoly/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release) will be automated.

## Linked issues

Partially addresses #3163.

### CODE REVIEW

File: .github/workflows/automate-changelog-PR.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    - uses: actions/checkout@v2
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.8
+    - name: Install changelog
+      run: pip install changelog-neuropoly
+    - name: Generate changelog
+      run: |
+        echo -e "MILESTONE=$(changelog neuropoly/spinalcordtoolbox --update --name ${{ github.event.inputs.fname_changelog }} | grep 'INFO Milestone' | head -1 | cut -d' ' -f3 | cut -d',' -f1)" >> ${GITHUB_ENV}
```
Comment (joshuacwnewton) [inline .github/workflows/automate-changelog-PR.yml:24]: This is a bit of trickery to get the milestone name from the stdout of `changelog`. Perhaps changelog could be updated to make this easier? Or, perhaps there's an easier way already -- suggestions appreciated!

File: .github/workflows/automate-changelog-PR.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        rm "${{ github.event.inputs.fname_changelog }}.bak"
+    - name: Create Pull Request
+      uses: peter-evans/create-pull-request@v3
+      with:
+        branch: "bot/v${{ env.MILESTONE }}"
+        commit-message  : "Update ${{ github.event.inputs.fname_changelog }} for ${{ env.MILESTONE }} release"
+        title: "Update ${{ github.event.inputs.fname_changelog }} for ${{ env.MILESTONE }} release"
+        milestone: "${{ env.MILESTONE }}"
+        body: "Automated step carried out as a part of [Creating a new release](https://github.com/neuropoly/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release) checklist."
+        delete-branch: true
```
Comment (joshuacwnewton) [inline .github/workflows/automate-changelog-PR.yml:34]: Also, one of the options that wasn't included here here is `reviewers:` -- we can automatically request reviews from the main SCT dev team if we want, too.

File: .github/workflows/automate-changelog-PR.yml (hunk 1 above)
Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:24]: IMO a better solution would be to explicitly specify the milestone name. There's an open feature request for this here: https://github.com/neuropoly/changelog/issues/15

Explicitly specifying the name would also avoid the issue described in https://github.com/neuropoly/changelog/issues/22.

Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:24]: Done in https://github.com/neuropoly/changelog/pull/23.

Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml]: Merged upstream, applied to this PR in d9a90f3.

**Comment by kousu** [comment]
This is very very cool. I didn't see this go by at all.

File: .github/workflows/automate-changelog-PR.yml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        rm "${{ github.event.inputs.fname_changelog }}.bak"
+    - name: Create pull request for updated changelog
+      uses: peter-evans/create-pull-request@v3
+      with:
+        path: "${{ github.event.repository.name }}"
+        branch: "bot/v${{ github.event.inputs.milestone_title }}"
+        commit-message  : "Update ${{ github.event.inputs.fname_changelog }} for ${{ github.event.inputs.milestone_title }} release"
+        title: "Update ${{ github.event.inputs.fname_changelog }} for ${{ github.event.inputs.milestone_title }} release"
+        milestone: "${{ github.event.inputs.milestone_title }}"
+        body: "Automated step carried out as a part of [Creating a new release](https://github.com/neuropoly/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release) checklist."
```
Comment (kousu) [inline .github/workflows/automate-changelog-PR.yml:47]: One of the examples you made with this, https://github.com/joshuacwnewton/test-github-actions/pull/16, looks like

2021-04-07-115012_781x460_scrot

I don't think you need to repeat "Automated", since Github already tags it with "bot", but my eye skipped over the title when looking at this. How about this wording:


```suggestion
        body: |
          Update ${{ github.event.inputs.fname_changelog }} for ${{ github.event.inputs.milestone_title }} release.
          
          Carried out as a part of the [release](https://github.com/neuropoly/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release) checklist."
```

Comment (kousu) [inline .github/workflows/automate-changelog-PR.yml:47]: Do you this it would be possible to *tag the Project*, so that CHANGES.md shows up in e.g. https://github.com/neuropoly/spinalcordtoolbox/projects/15 ?

File: .github/workflows/automate-changelog-PR.yml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        pip install ./changelog
+    - name: Checkout '${{ github.event.repository.name }}'
+      uses: actions/checkout@v2
+      with:
+        path: ${{ github.event.repository.name }}
+    - name: Update '${{ github.event.inputs.fname_changelog }}' for '${{ github.event.repository.name }}'
+      run: |
+        cd ${{ github.event.repository.name }}
+        changelog neuropoly/spinalcordtoolbox --update --name "${{ github.event.inputs.fname_changelog }}" --milestone ${{ github.event.inputs.milestone_title }}
+        rm "${{ github.event.inputs.fname_changelog }}.bak"
```
Comment (kousu) [inline .github/workflows/automate-changelog-PR.yml:38]: Huh, why? We have git to serve as a backup system. Why is changelog duplicating that work?

https://github.com/neuropoly/changelog/pull/27

**Review by kousu — APPROVED** [review]
This is cool. If you wanna merge it go for it.

File: .github/workflows/automate-changelog-PR.yml (hunk 4 above)
Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:36]: Good catch. :smile:

File: .github/workflows/automate-changelog-PR.yml (hunk 3 above)
Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:45]: I wish this were possible. I think the core problem here is that GitHub has poor integration between milestones and project boards.

What I _really_ wish I could do is link a project board to a milestone. I want to be able to have any issues/PRs tagged with a milestone to be automatically added to the corresponding project board.

(Without this, we're running into friction where Julien will add a milestone to an issue/PR, then wonder why it's not also on the project board... we're kind of using 2 different methods for release tracking...)

File: .github/workflows/automate-changelog-PR.yml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      with:
+        python-version: 3.8
+    - name: Checkout 'changelog'
+      uses: actions/checkout@v2
+      with:
+        repository: neuropoly/changelog
+        path: changelog
+    - name: Install 'changelog'
+      run: |
+        pip install ./changelog
```
[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:27]: Aha, good thinking.

File: .github/workflows/automate-changelog-PR.yml (hunk 3 above)
Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:40]: Ohhh... wait just one minute. https://github.com/marketplace/actions/move-labeled-or-milestoned-issue-to-a-specific-project-column

Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:40]: I think I could create a GH actions workflow that triggers when an issue/PR is milestoned, then adds it to a project board of the same name. (Which would solve both issues described above.)

Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:40]: Opened https://github.com/neuropoly/spinalcordtoolbox/issues/3338. Since this is a larger issue with our use of project boards, I'm going to mark this as "resolved" for now.

File: .github/workflows/automate-changelog-PR.yml (hunk 4 above)
Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml:31]: I should probably change this to `rm -f` that way the workflow doesn't fail abruptly if https://github.com/neuropoly/changelog/pull/27 is addressed.

**Comment by joshuacwnewton** [comment]
Thanks for the feedback, @kousu! I've applied your changes. Now I'll just test this over on the other repo and see how things go.

File: .github/workflows/automate-changelog-PR.yml (hunk 2 above)
Comment (joshuacwnewton) [reply .github/workflows/automate-changelog-PR.yml]: I think this won't be viable until #3324 is addressed.

**Comment by joshuacwnewton** [comment]
I've tested the changes here: https://github.com/joshuacwnewton/test-github-actions/pull/17

Everything looks good to me, minus the one typo, but this was fixed in 0203ba1. So, I'm going to merge.

**Comment by kousu** [comment]
Great!
